src/control/lib/hardware/hwloc/bindings.go

```diff
@@ -454,6 +476,56 @@ func (o *object) pciAddr() (*hardware.PCIAddress, error) {
 		pciDevAttr.dev, pciDevAttr._func))
 }
 
+func (o *object) blockDevice() (*hardware.BlockDevice, error) {
+	o.topo.RLock()
+	defer o.topo.RUnlock()
+
+	devType, err := o.osDevType()
+	if err != nil {
+		return nil, err
+	}
+	if devType != osDevTypeBlock {
+		return nil, errors.Errorf("device %q is not a block device", o.name())
+	}
+
+	bd := &hardware.BlockDevice{
+		Name: o.name(),
+		Type: o.objSubTypeString(),
+	}
+	// https://www.open-mpi.org/projects/hwloc/doc/v2.7.0/a00365.php (OS Device Information)
+	infoKeys := map[string]unsafe.Pointer{
+		"LinuxDeviceID": unsafe.Pointer(&bd.LinuxDeviceID),
+		"Vendor":        unsafe.Pointer(&bd.Vendor),
+		"Model":         unsafe.Pointer(&bd.Model),
+		"Revision":      unsafe.Pointer(&bd.Revision),
+		"SerialNumber":  unsafe.Pointer(&bd.SerialNumber),
+		"Size":          unsafe.Pointer(&bd.Size),
+		"SectorSize":    unsafe.Pointer(&bd.SectorSize),
+	}
```

Comment:
I can appreciate this efficiency, but at the same time I feel like it obfuscates things quite a bit. Casting the fields of a Go struct into unsafe.Pointers so we can fill them out by casting to an appropriate type later seems unnecessarily dangerous, too.

Reply:
What would you propose as an alternative? It doesn't seem particularly dangerous to me, given that we're allocating the struct on the stack and using it in the same method, so it's easy enough to look at the whole thing at once. I probably wouldn't do something like this if things were further apart, though.

Reply:
I was thinking you could assign values in a more conventional way (directly assigning a Go string or an integer). My concern was about the type handling--the compiler protects us if we use the type directly, vs. potentially clobbering memory and creating hard-to-analyze runtime bugs. Not sure of the likelihood that we expand this struct in the future, or change types of fields, but if that happens we will need to remember to update here and/or in the loop.

Reply:
after looking at this little I think it's a good approach to populating the struct and I don't see a cleaner way to do it

Reply:
> My concern was about the type handling--the compiler protects us if we use the type directly, vs. potentially clobbering memory and creating hard-to-analyze runtime bugs. 

I don't think it's quite as unsafe as you might think... We're not discarding the type safety (as we might be with `interface{}`). In the assignments in the switch cases, we're still asserting concrete types. The compiler is still working for us here, e.g. if I try to arbitrarily assign a `uint64` to a dereferenced string pointer (`*(*string)`), the compiler will complain: `cannot use size (variable of type uint64) as type string in assignment`.

Reply:
Good to know. Thanks.

Reply:
Thinking about it a bit more, I decided that it's probably best to use two maps and avoid any possibility of ambiguity. We still get the efficiency and simplicity of taking references to the struct fields, but guarantee that a change in the type of a struct field will result in a compile error instead of a runtime error.